### PR TITLE
[Compression] add basic distiller

### PR DESCRIPTION
**Pull request by J-shang**
### Description ###

please review commit [e335f37](https://github.com/microsoft/nni/pull/5397/commits/e335f3788783875a274ab6b1e71c5e60047158f4)

Add two teacher model based distillation algo, `DynamicLayerwiseDistiller`, `Adaptive1dLayerwiseDistiller`.

#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [ ] test case
  - [ ] doc

### CODE REVIEW

File: nni/contrib/compression/base/wrapper.py [hunk 1]
```diff
@@ -186,6 +186,13 @@ def _transfer_input(self, *args, **kwargs) -> Tuple:
         pos_args = args[:pos_args_num]
         if len(pos_args) < pos_args_num:
             pos_args += tuple(kwargs.pop(k) for k in self._input_args_spec.args[len(pos_args) + 1:])
+            default_values = {}
+            if self._input_args_spec.defaults:
```
Comment (ultmaster) [inline nni/contrib/compression/base/wrapper.py:190]: Add comments about what types of inputs you are trying to deal with here.

File: nni/contrib/compression/distillation/basic_distiller.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+        self._fusion_compress(max_steps, max_epochs)
+
+    def _fuse_preprocess(self, evaluator: Evaluator):
+        self._register_loss_patch(evaluator)
+
+    def _fuse_postprocess(self, evaluator: Evaluator):
+        pass
+
+
+class DynamicLayerwiseDistiller(TeacherModelBasedDistiller):
```
Comment (ultmaster) [inline nni/contrib/compression/distillation/basic_distiller.py:142]: What is `DynamicLayerwiseDistiller`?

File: nni/contrib/compression/distillation/basic_distiller.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+        for _, ts in self._target_spaces.items():
+            for _, target_space in ts.items():
+                target_space.clean()
+        for _, ts in self._teacher_target_spaces.items():
+            for _, target_space in ts.items():
+                target_space.clean()
+        return distill_loss
+
+
+class Adaptive1dLayerwiseDistiller(TeacherModelBasedDistiller):
```
Comment (ultmaster) [inline nni/contrib/compression/distillation/basic_distiller.py:208]: Is this parallel with `DynamicLayerwiseDistiller`?

**Review by ultmaster — APPROVED** [review]

File: nni/contrib/compression/distillation/basic_distiller.py (hunk 2 above)
Comment (J-shang) [reply nni/contrib/compression/distillation/basic_distiller.py:142]: Each student layer corresponds to a set of teacher layers, during distillation, the student layer will compute a set of distil losses with each teacher layer in its teacher set, then choose the smallest loss in the loss set as this student layer distil loss.

I will add comments for this distiller.

File: nni/contrib/compression/distillation/basic_distiller.py [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+class TeacherModelBasedDistiller(Distiller):
+    """
+    The base class that the distiller need a teacher model.
+
+    Parameters
+    ----------
+    model
+        The student model to be distilled.
+    config_list
+        A list of dict, each dict configure which module need to be distilled, and how to distil.
```
Comment (QuanluZhang) [inline nni/contrib/compression/distillation/basic_distiller.py:31]: distil -> distill

File: nni/contrib/compression/distillation/basic_distiller.py [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+        Please refer :doc:`Compression Config Specification </compression/compression_config_list>` for more information.
+    evaluator
+        Please refer TODO.
```
Comment (QuanluZhang) [inline nni/contrib/compression/distillation/basic_distiller.py:34]: still TODO here?

Comment (J-shang) [reply nni/contrib/compression/distillation/basic_distiller.py:34]: yes, will update for all compressor after evaluator doc done...

**Review by QuanluZhang — APPROVED** [review]

**Pull request merged**